### PR TITLE
Enable compiler warnings in the CI

### LINKED ISSUE #180: Enable compiler warnings in the CI
They have been a number of PR to fix warnings in the last couple weeks (e.g., #173, #178, #179).
We should at the very least enable compiler warnings in the CI build and then consider either passing `-Werror` or try to setup a Jenkins plugin, probably [this one](https://jenkins.io/doc/pipeline/steps/warnings-ng/).

### PR DESCRIPTION
**Pull request by dalg24**
Fix #180 

TODO: Decide how to report warnings.

### CODE REVIEW

File: .jenkins [hunk 1]
```diff
@@ -23,7 +23,19 @@ pipeline {
...
                         sh 'rm -rf build && mkdir -p build'
                         dir('build') {
-                            sh 'cmake -D CMAKE_INSTALL_PREFIX=$ARBORX_DIR -D CMAKE_BUILD_TYPE=Debug -D CMAKE_CXX_COMPILER_LAUNCHER=ccache -D CMAKE_CXX_COMPILER=$KOKKOS_DIR/bin/nvcc_wrapper -D CMAKE_PREFIX_PATH="$KOKKOS_DIR;$BOOST_DIR;$BENCHMARK_DIR" -D ARBORX_ENABLE_MPI=ON -D ARBORX_USE_CUDA_AWARE_MPI=ON -D MPIEXEC_PREFLAGS="--allow-run-as-root" ..'
+                            sh '''
+                                cmake \
+                                    -D CMAKE_INSTALL_PREFIX=$ARBORX_DIR \
+                                    -D CMAKE_BUILD_TYPE=Debug \
+                                    -D CMAKE_CXX_COMPILER_LAUNCHER=ccache \
+                                    -D CMAKE_CXX_COMPILER=$KOKKOS_DIR/bin/nvcc_wrapper \
+                                    -D CMAKE_CXX_WARNINGS="-Wpedantic -Wall -Wextra -Wshadow" \
```
Comment (masterleinad) [inline .jenkins]:
```suggestion
                                    -D CMAKE_CXX_FLAGS="-Wpedantic -Wall -Wextra -Wshadow" \
```

**Comment by masterleinad** [comment]
This seems to be useful for the `clang` builds, but `nvcc_wrapper` is way too noisy. It seems that `nvcc_wrapper` does not see the `thrust` headers as system include files while `clang` does.

**Comment by aprokop** [comment]
> Anyone willing to give a shot to the Jenkins plugin?

What does it mean?

**Comment by masterleinad** [comment]
It seems that there is no way to get these reports displayed in `BlueOcean` (https://issues.jenkins-ci.org/browse/JENKINS-55359).
For me, a solution using `Werror` and running `clang-tidy` with warnings as errors appears to be nicer. Of course, we need to see if we can filter the spurious warnings anyway.

**Comment by dalg24** [comment]
> It seems that there is no way to get these reports displayed in `BlueOcean` (https://issues.jenkins-ci.org/browse/JENKINS-55359).
> For me, a solution using `Werror` and running `clang-tidy` with warnings as errors appears to be nicer. Of course, we need to see if we can filter the spurious warnings anyway.

The solution you suggest would imply the unit tests and other checks do not run.  The display (that is still accessible on the classic Jenkins interface BTW) is one thing but we can also setup a threshold that make the CI report failure after the fact if it exceeded.

File: .jenkins [hunk 2]
```diff
@@ -145,4 +190,16 @@ pipeline {
...
         }
     }
+    post {
+        always {
+            node('docker') {
+                recordIssues(
+                    enabledForFailure: true,
+                    tools: [cmake(), gcc(), clang(), clangTidy()],
+                    qualityGates: [[threshold: 1, type: 'TOTAL', unstable: true]],
+                    filters: [excludeFile('/usr/local/cuda.*')]
```
Comment (dalg24) [inline .jenkins]: This is not ideal but I am not sure how to to make it a system header path with CMake.  The path is implicitly added by NVCC.

**Comment by dalg24** [comment]
This is ready for review/merge.

File: src/details/ArborX_DetailsTreeVisualization.hpp [hunk 3]
```diff
@@ -197,6 +197,9 @@ struct TreeVisualization
     // warnings about calling a __host__ function from a __host__
     // __device__ function emitted by nvcc.
 #if defined(__CUDA_ARCH__)
+    (void)tree;
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeVisualization.hpp]: Use `std::ignore`.

**Pull request merged**